Allow for hdfs and gcs URI's to be passed to GenomicsDB

nalinigans opened the pull request:
Currently, only Posix filesystem paths can be passed as workspaces and arrays to GenomicsDB via GenomicsDBImport and SelectVariants. This PR will allow for hdfs and gcs (and emrfs/s3) URIs to be supported as well. 
Examples
   ```
./gatk GenomicsDBImport -V /vcfs/sample.vcf.gz --genomicsdb-workspace-path hdfs://master:9000/gdb_ws  -L 1:500-10000
   ./gatk GenomicsDBImport -V /vcfs/sample.vcf.gz --genomicsdb-workspace-path gs://my_bucket/gdb_ws  -L 1:500-10000
```
```
  ./gatk SelectVariants -V gendb.hdfs://master:9000/gdb_ws -R hs37d5.fa -O out.vcf
 ./gatk SelectVariants -V gendb.gs://my_bucket/gdb_ws -R hs37d5.fa -O out.vcf
```
GenomicsDB supports GCS via the [Cloud Storage Connector](https://cloud.google.com/dataproc/docs/concepts/connectors/cloud-storage). Set environment variable GOOGLE_APPLICATION_CREDENTIALS to point to the GCS Service Account json file.

kgururaj commented:
I think you should use version 0.10.0-proto-3.0.0-beta-1+d392491bafcac337 for GenomicsDB

droazen commented on src/main/java/org/broadinstitute/hellbender/engine/FeatureDataSource.java:
```diff
@@ -369,29 +370,13 @@ public static boolean isGenomicsDBPath(final String path) {
-                    GenomicsDBConstants.DEFAULT_VIDMAP_FILE_NAME + "," + GenomicsDBConstants.DEFAULT_CALLSETMAP_FILE_NAME + "," +
-                    GenomicsDBConstants.DEFAULT_VCFHEADER_FILE_NAME + ") could not be read from GenomicsDB workspace " + workspace.getAbsolutePath(), e);
-        }
+        final String workspace = path.replace(GENOMIC_DB_URI_SCHEME, "");
```
Regex should have an anchor tying it to the start of the String.

nalinigans replied:
Agreed, we are using Pattern/Matcher to check valid gendb names, e.g. gendb://file, gendb.hdfs://node:portnum/file and gendb.gs://bucket/file.

droazen commented on src/main/java/org/broadinstitute/hellbender/engine/FeatureDataSource.java:
```diff
@@ -369,29 +370,13 @@ public static boolean isGenomicsDBPath(final String path) {
+        final String workspace = path.replace(GENOMIC_DB_URI_SCHEME, "");
+        final String callsetJson = BucketUtils.appendPathToDir(workspace, GenomicsDBConstants.DEFAULT_CALLSETMAP_FILE_NAME);
+        final String vidmapJson = BucketUtils.appendPathToDir(workspace, GenomicsDBConstants.DEFAULT_VIDMAP_FILE_NAME);
+        final String vcfHeader = BucketUtils.appendPathToDir(workspace, GenomicsDBConstants.DEFAULT_VCFHEADER_FILE_NAME);
```
This method would be more at home in `IOUtils`

nalinigans replied:
Are you talking about the BucketUtils.appendPathToDir() method? I was thinking it would belong  in IOUtils too, but was following the example of BucketUtils.makeFilePathAbsolute(). I will move appendPathToDir() to IOUtils however. Thanks.

nalinigans replied:
Moved appendPathToDir to IOUtils.

droazen commented on src/main/java/org/broadinstitute/hellbender/utils/gcs/BucketUtils.java:
```diff
@@ -81,6 +81,27 @@ public static String makeFilePathAbsolute(String path){
+     * @param path the path relative to dir
+     * @return the appended path as a String.
+     */
+    public static String appendPathToDir(String dir, String path) {
```
Can you convert the String to a `Path`, and use `resolve()` instead of adding this method?

nalinigans replied:
Sure, will do.

droazen commented on src/main/java/org/broadinstitute/hellbender/engine/FeatureDataSource.java:
```diff
@@ -400,21 +385,20 @@ public static boolean isGenomicsDBPath(final String path) {
-                        .setVidMappingFile(vidmapJson.getAbsolutePath())
-                        .setCallsetMappingFile(callsetJson.getAbsolutePath())
-                        .setVcfHeaderFilename(vcfHeader.getAbsolutePath())
+                        .setVidMappingFile(vidmapJson)
```
Are these guaranteed to be absolute at this point? (the previous code enforced this)

nalinigans replied:
I have made sure that all these paths are absolute in the changed code where we are allowing for gendb://, gendb.hdfs:// and gendb.gs:// schemes.

droazen commented on src/main/java/org/broadinstitute/hellbender/tools/genomicsdb/GenomicsDBImport.java:
```diff
@@ -649,27 +650,12 @@ public Object onTraversalSuccess() {
-        }
+    private String overwriteOrCreateWorkspace() {
+	String workspaceDir = BucketUtils.makeFilePathAbsolute(workspace);
+	if (GenomicsDBUtils.createTileDBWorkspace(workspaceDir, overwriteExistingWorkspace) < 0) {
```
If the workspace already exists and `overExistingWorkspace` is false, we want an `UnableToCreateGenomicsDBWorkspace` exception to be thrown, as in the previous implementation.

droazen commented on src/main/java/org/broadinstitute/hellbender/tools/genomicsdb/GenomicsDBImport.java:
```diff
@@ -680,13 +666,6 @@ private File overwriteOrCreateWorkspace() {
     }
 
-    private static void checkIfValidWorkspace(final File workspaceDir) {
-        final File tempFile = new File(workspaceDir.getAbsolutePath() + "/__tiledb_workspace.tdb");
```
Is this check performed internally now?

nalinigans replied:
Yes, this check is performed internally in GenomicsDB now.

droazen commented on src/test/java/org/broadinstitute/hellbender/utils/gcs/BucketUtilsTest.java:
```diff
@@ -57,6 +58,19 @@ public void testIsFileURL(){
+	Assert.assertEquals(BucketUtils.appendPathToDir("hdfs://namenode:9000/dir", "file"), "hdfs://namenode:9000/dir/file");
+	Assert.assertEquals(BucketUtils.appendPathToDir("gs://abucket/dir", "file"), "gs://abucket/dir/file");
+    }
+
```
I think we need some tests covering the new functionality introduced by this PR (ie., GCS support). Let us know if you need help setting this up -- there are examples of tests that access data in GCS buckets in `GCSNioIntegrationTest`

nalinigans replied:
@droazen, yes, I need some help setting up tests for using hdfs and gs URIs with GenomicsDB. I am taking a look at `GCSNioIntegrationTest`, but still not sure on setting up a test bucket with folder, etc. on GCS and getting authentication setup via GOOGLE_APPLICATIONS_CREDENTIALS for GenomicsDB.

droazen commented on src/main/java/org/broadinstitute/hellbender/tools/genomicsdb/GenomicsDBImport.java:
```diff
@@ -529,9 +530,9 @@ private ImportConfig createImportConfig(final int batchSize) {
-        importConfig.setOutputCallsetmapJsonFile(callsetMapJSONFile.getAbsolutePath());
-        importConfig.setOutputVidmapJsonFile(vidMapJSONFile.getAbsolutePath());
-        importConfig.setOutputVcfHeaderFile(vcfHeaderFile.getAbsolutePath());
+        importConfig.setOutputCallsetmapJsonFile(callsetMapJSONFile);
```
Are these guaranteed to be absolute at this point? (the previous implementation guaranteed this)

nalinigans replied:
Yes. These are generated with respect to the workspace directory that is made absolute - see line 655 : String workspaceDir = BucketUtils.makeFilePathAbsolute(workspace);

droazen commented on src/main/java/org/broadinstitute/hellbender/engine/FeatureDataSource.java:
```diff
@@ -369,29 +370,13 @@ public static boolean isGenomicsDBPath(final String path) {
             throw new IllegalArgumentException("Trying to create a GenomicsDBReader from a non-GenomicsDB input");
         }
 
```
Instead of `gendb://gs://` as the URI convention, can you introduce a new URI scheme for "gendb on gcs"? Eg., `gendbgs://` (and still also accept `gendb://` for local gendb instances). 

This way, the URIs will at least be legal...

nalinigans replied:
For the URI schemes, would you consider gendb://, gendb.gs:// gendb.hdfs://, etc.?
According to [wikipedia](https://en.wikipedia.org/wiki/Uniform_Resource_Identifier), these are valid schemes - A non-empty scheme component followed by a colon (:), consisting of a sequence of characters beginning with a letter and followed by any combination of letters, digits, plus (+), period (.), or hyphen (-).

droazen replied:
If a period in the scheme is legal, then these would be fine. I'd check to make sure that Java's URI, URL, and `java.nio.file.Path` classes accept such schemes, though.

nalinigans replied:
Yes, URI and Path recognize period in their schemes.

nalinigans commented:
@droazen, I think I have pushed most of the changes requested -

* Moved out `appendPathToDir` from BucketUtils to IOUtils
* `appendPathToDir` now uses Path.resolve() to append a given path to dir
* If a workspace already exists and `overExistingWorkspace` is false, a `UnableToCreateGenomicsDBWorkspace` exception is thrown while creating a GenomicsDB workspace.
* Introduced `gendb.hdfs:` and `gendb.gs:` URI schemes in addition to the existing `gendb:` scheme for identifying Cloud paths in GenomicsDB with unit testing for these new schemes.
* Added unit tests to test out writing out GenomicsDB workspace/arrays to GCS and then reading/querying the same GenomicsDB instance from GCS with validation.

droazen commented on src/test/java/org/broadinstitute/hellbender/tools/genomicsdb/GenomicsDBImportIntegrationTest.java:
```diff
@@ -779,4 +780,56 @@ public void testYouCantWriteIntoAnExistingDirectory(){
         writeToGenomicsDB(LOCAL_GVCFS, INTERVAL, workspace, 0, false, 0, 1);
     }
+
+    private void cleanupGCSFolder(String path) {
```
Instead of implementing this cleanup method here, could you add a `BucketUtils.deleteRecursively(Path)` method, and use the slightly cleaner approach below:

```
public static void deleteRecursively(final Path rootPath) {
    final List<Path> pathsToDelete = Files.walk(rootPath).sorted(Comparator.reverseOrder()).collect(Collectors.toList());
    for (Path path : pathsToDelete) {
        Files.deleteIfExists(path);
    }
}
```

Then modify `BucketUtils.deleteOnExit()` to call into `deleteRecursively()` instead of `deleteFile()`. The advantage of this approach is that if you switch to calling `BucketUtils.getTempFilePath()` in your test methods below, the returned temp paths will be automatically scheduled for deletion on JVM exit, and you can get rid of the finally blocks and the `cleanupGCSFolder()` calls in your test code.

droazen commented on src/test/java/org/broadinstitute/hellbender/tools/genomicsdb/GenomicsDBImportIntegrationTest.java:
```diff
@@ -779,4 +780,56 @@ public void testYouCantWriteIntoAnExistingDirectory(){
+    public void testWriteToAndQueryFromGCS() throws IOException {
+        String workspace = BucketUtils.randomRemotePath(getGCPTestInputPath(), "","");
+        try {
+            Assert.assertNotNull(getGoogleServiceAccountKeyPath());
```
You don't need this check (or the `UserException` catch block) in your test -- tests that are part of the "bucket" group only get run in travis when GOOGLE_APPLICATION_CREDENTIALS is already set.

nalinigans commented on src/main/java/org/broadinstitute/hellbender/engine/FeatureDataSource.java:
```diff
@@ -287,9 +293,61 @@ public FeatureDataSource(final FeatureInput<T> featureInput, final int queryLook
+     *             gendb.s3://my_bucket/my_folder
+     * @return GenomicsDB acceptable path or null
+     */
+    public static String getGenomicsDBPath(final String path) {
```
Agreed. I was just following the existing pattern. Would you rather have a GenomicsDBUtils class in the org.broadinstitute.hellbender.tools.genomicsdb package instead? That might be cleaner.

nalinigans commented on src/main/java/org/broadinstitute/hellbender/engine/FeatureDataSource.java:
```diff
@@ -287,9 +293,61 @@ public FeatureDataSource(final FeatureInput<T> featureInput, final int queryLook
+    }
+
+    public static String getAbsolutePathWithGenDBScheme(final String path) {
+        String gendb_path = FeatureDataSource.getGenomicsDBAbsolutePath(path);
```
Meant to have them all camel-case, I guess my C/C++ roots creep in.

nalinigans commented on src/main/java/org/broadinstitute/hellbender/tools/genomicsdb/GenomicsDBImport.java:
```diff
@@ -650,26 +648,20 @@ public Object onTraversalSuccess() {
+        //   returnCode = -1 : path was not a directory
+        //   returnCode = -2 : failed to create workspace
+        //   returnCode = 1 : existing directory, nothing changed
+        int returnCode = GenomicsDBUtils.createTileDBWorkspace(workspaceDir, overwriteExistingWorkspace);
```
The value of `overwriteExitingWorkspace` does affect the return value of the method.
If `overwriteExistingWorkspace` is `true` and the workspace exists, it is deleted first. The return value is `0` if the delete operation and the create workspace operation are both successful.
If `overwriteExistingWorkspace` is `false` and the workspace exists, the return value is `1` and the existing directory is left intact. I have added comments to make this more clear.

droazen commented on src/test/java/org/broadinstitute/hellbender/utils/io/IOUtilsUnitTest.java:
```diff
@@ -205,6 +205,25 @@ public void testGetAbsolutePathWithoutFileProtocol(final String uriString, final
+        Assert.assertEquals(IOUtils.appendPathToDir("/path/to/dir", "anotherdir/file"), "/path/to/dir/anotherdir/file");
+
+        // hdfs: URI
+        Path tempPath = IOUtils.getPath(MiniClusterUtils.getWorkingDir(MiniClusterUtils.getMiniCluster()).toUri().toString());
```
Need to stop the cluster when done.

Actually there's no good reason to use a `MiniCluster` here -- just construct an appropriate String instead.

droazen commented on src/test/java/org/broadinstitute/hellbender/utils/io/IOUtilsUnitTest.java:
```diff
@@ -370,4 +389,42 @@ public void testIsHDF5File(final String filePath, final boolean expected) {
+    public void testGenomicsDBPathParsing(String path, String expectedPath, String gendbExpectedAbsolutePath, boolean expectedComparison) {
+        Assert.assertEquals(IOUtils.getGenomicsDBPath(path), expectedPath, "Got 1 "+IOUtils.getGenomicsDBPath(path));
+        Assert.assertEquals(IOUtils.getAbsolutePathWithGenDBScheme(path), gendbExpectedAbsolutePath);
+        Assert.assertEquals(IOUtils.isGenomicsDBPath(path), expectedComparison, "Got 3 " + IOUtils.isGenomicsDBPath(path));
```
Assertion failure messages here could be better worded.

droazen commented:
I've opened https://github.com/broadinstitute/gatk/pull/5197 with the last few comments addressed. We'll see if the travis cloud tests pass (since they were never actually run on this PR), and if they do pass we can merge.

nalinigans commented:
@droazen, I have pushed a debugging test. It just prints out all the keys and not private and not id values in the service json pointed by GOOGLE_APPLICATION_CREDENTIALS env. Would it be possible to accept this into the  nalinigans_genomicsdb_uri_support branch, so I can browse through the stdout for that test on Travis? By the way, the failures in the build seem to be unrelated to my change. Thanks.

droazen commented:
Closing this PR in favor of https://github.com/broadinstitute/gatk/pull/5197
